### ISSUE TITLE
Array slicing breaks csr_AnnotationMatrix objects

### ISSUE THREAD

**Issue by jason-fries** [HazyResearch/snorkel#582]
I'd like to take a subset of candidate instances to train the generative model and then save those marginals for use in the discriminative model. 

I thought this could be done just taking a subset slice of the label matrix.  However, when I use  slice notation on a csr_AnnotationMatrix object to generate my model input, I run into errors when attempting to save the marginals later on in the program. For example:

```
L_train = labeler.load_matrix(session, split=0)
print L_train.get_candidate(session,0)
```
will correctly access and the mapped candidate object. 
However, if we take a slice of the label matrix, e.g., 
```
L_train_subset = L_train[0:n,...]
print L_train_subset.get_candidate(session,0)
```
the `get_candidate` function breaks with the following error
```
File "/home/user/code/dd-patient-notes/snorkel/snorkel/annotations.py", line 39, in get_candidate
    return session.query(Candidate).filter(Candidate.id == self.row_index[i]).one()
TypeError: 'NoneType' object has no attribute '__getitem__'
```
everything works otherwise in terms of training, just the mapping between candidate and row index seems to break.

**Comment by ajratner** [comment]
Hey Jason- thanks for pointing this out!  We just need to add a method for
array slicing so this is handled properly and preserves the index

**Comment by bhancock8** [comment]
I'd just like to second that this will be a very useful feature. :) For Pretty Ricky, there are cases where I'd like to remove certain LFs (columns from the label matrix) after I see that they label identically to another LF or are otherwise less than helpful (label all true or all false).

**Comment by ajratner** [comment]
See PR #688.  I can't figure out how to / what subfunction handles list/array slicing, but normal slices (e.g. `a:b`) work now... maybe @bhancock8 or @jason-fries could pick up from where I left off and give a shot?  Or just merge in what I have?

**Comment by jason-fries** [comment]
I tried implementing this a couple months ago, but ended up abandoning my attempts to override the slice operator. I don't think it's actually possible to implement a pure Python override that fully supports Numpy's slice syntax.  I think it's better to either:

1) Allow `load_matrix` and `apply` functions to operate over arbitrary candidate sets (see Issue #686 ), enumerated via a query. I just hacked together an implemented for this on the `ehr` branch.

2) Have all learning/labeling code accept standard sparse matrices and assume the correct ordering. Then we just use slice operators as normal.
